x-pack/plugin/gpu/src/main/java/org/elasticsearch/xpack/gpu/codec/ES92GpuHnswVectorsWriter.java

```diff
@@ -448,105 +446,205 @@ public NodesIterator getNodesOnLevel(int level) {
+        } catch (Throwable t) {
+            throw new IOException("Failed to merge GPU index: ", t);
         }
-        try (IndexInput in = mergeState.segmentInfo.dir.openInput(tempRawVectorsFileName, IOContext.DEFAULT)) {
-            var input = FilterIndexInput.unwrapOnlyTest(in);
-
-            if (numVectors >= MIN_NUM_VECTORS_FOR_GPU_BUILD) {
-                if (input instanceof MemorySegmentAccessInput memorySegmentAccessInput) {
-                    // Direct access to mmapped file
-
-                    try (
-                        var dataset = DatasetUtils.getInstance()
-                            .fromInput(memorySegmentAccessInput, numVectors, fieldInfo.getVectorDimension(), dataType);
-                        var resourcesHolder = new ResourcesHolder(
-                            cuVSResourceManager,
-                            cuVSResourceManager.acquire(numVectors, fieldInfo.getVectorDimension(), dataType)
-                        )
-                    ) {
+    }
+
+    private void mergeByteVectorField(
+        FieldInfo fieldInfo,
+        MergeState mergeState,
+        RandomVectorScorerSupplier randomScorerSupplier,
+        int numVectors
+    ) throws IOException, InterruptedException {
+        var vectorValues = randomScorerSupplier == null
+            ? null
+            : VectorsFormatReflectionUtils.getByteScoringSupplierVectorOrNull(randomScorerSupplier);
+        if (vectorValues != null) {
+            IndexInput slice = vectorValues.getSlice();
+            var input = FilterIndexInput.unwrapOnlyTest(slice);
+            if (input instanceof MemorySegmentAccessInput memorySegmentAccessInput) {
+                // Direct access to mmapped file
+                // TODO: strides!!
+                // for int8_hnsw, the raw vector data has extra 4-byte at the end of each vector to encode a correction constant
+                int rowStride = fieldInfo.getVectorDimension() + 4;
+                try (
+                    var dataset = DatasetUtils.getInstance()
+                        .fromInput(memorySegmentAccessInput, numVectors, fieldInfo.getVectorDimension(), rowStride, -1, dataType);
+                    var resourcesHolder = new ResourcesHolder(
+                        cuVSResourceManager,
+                        cuVSResourceManager.acquire(numVectors, fieldInfo.getVectorDimension(), dataType)
+                    );
+                    // Explicitly copy the dataset to GPU memory. The current (25.10) CAGRA index implementation has
+                    // problems with strides; the explicit copy removes the stride while copying.
+                    // Note that this is _not_ an additional copy: input data needs to be moved to GPU memory anyway,
+                    // we are just doing it explicitly instead of relying on CagraIndex#build to do it.
+                    var deviceDataSet = dataset.toDevice(resourcesHolder.resources())
```

Comment:
Nice workaround, so you also confirmed that strides don't work properly with Cagra index implementation?